src/connector/src/sink/mqtt.rs

```diff
@@ -0,0 +1,254 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use core::fmt::Debug;
+use std::collections::HashMap;
+use std::sync::atomic::AtomicBool;
+use std::sync::Arc;
+
+use anyhow::{anyhow, Context as _};
+use risingwave_common::array::StreamChunk;
+use risingwave_common::catalog::Schema;
+use rumqttc::v5::mqttbytes::QoS;
+use rumqttc::v5::ConnectionError;
+use serde_derive::Deserialize;
+use serde_with::serde_as;
+use thiserror_ext::AsReport;
+use tokio_retry::strategy::{jitter, ExponentialBackoff};
+use tokio_retry::Retry;
+use with_options::WithOptions;
+
+use super::encoder::{DateHandlingMode, TimeHandlingMode, TimestamptzHandlingMode};
+use super::utils::chunk_to_json;
+use super::{DummySinkCommitCoordinator, SinkWriterParam};
+use crate::common::MqttCommon;
+use crate::sink::catalog::desc::SinkDesc;
+use crate::sink::encoder::{JsonEncoder, TimestampHandlingMode};
+use crate::sink::log_store::DeliveryFutureManagerAddFuture;
+use crate::sink::writer::{
+    AsyncTruncateLogSinkerOf, AsyncTruncateSinkWriter, AsyncTruncateSinkWriterExt,
+};
+use crate::sink::{Result, Sink, SinkError, SinkParam, SINK_TYPE_APPEND_ONLY};
+use crate::{deserialize_bool_from_string, deserialize_u32_from_string};
+
+pub const MQTT_SINK: &str = "mqtt";
+
+#[serde_as]
+#[derive(Clone, Debug, Deserialize, WithOptions)]
+pub struct MqttConfig {
+    #[serde(flatten)]
+    pub common: MqttCommon,
+
+    // 0 - AtLeastOnce, 1 - AtMostOnce, 2 - ExactlyOnce
+    #[serde(default, deserialize_with = "deserialize_u32_from_string")]
+    pub qos: u32,
```

Comment:
Recommend using dedicated types here and config via string. For example:

https://github.com/risingwavelabs/risingwave/blob/bb7d69d394285140da7fc7539cae64599387e13b/src/connector/src/sink/kafka.rs#L142-L145